pkg/scheduler/podtask/registry_test.go

```diff
@@ -0,0 +1,304 @@
+	another_b := b.Clone()
+	another_b.ID = "another-task-id"
+	another_b_clone, err := registry.Register(another_b, nil)
+	assert.Error(err)
+	assert.Nil(another_b_clone)
+
+	// unregister a task
+	registry.Unregister(b)
+
+	tasks = registry.List(func(t *T) bool { return true })
+	assert.Len(tasks, 1)
+	assert.Contains(tasks, a)
+
+	// unregister a task not registered
+	unregistered_task, _ := fakePodTask("unregistered-task")
+	registry.Unregister(unregistered_task)
+}
+
+func fakeStatusUpdate(taskId string, state mesos.TaskState) *mesos.TaskStatus {
+	status := mesosutil.NewTaskStatus(mesosutil.NewTaskID(taskId), state)
+	status.Data = []byte("{}") // empty json
+	masterSource := mesos.TaskStatus_SOURCE_MASTER
+	status.Source = &masterSource
+	return status
+}
+
+func TestInMemoryRegistry_State(t *testing.T) {
+	assert := assert.New(t)
+
+	registry := NewInMemoryRegistry()
+
+	// add a task
+	a, _ := fakePodTask("a")
+	a_clone, err := registry.Register(a, nil)
+	assert.NoError(err)
+	assert.Equal(a.State, a_clone.State)
+
+	// update the status
+	assert.Equal(a_clone.State, StatePending)
+	a_clone, state := registry.UpdateStatus(fakeStatusUpdate(a.ID, mesos.TaskState_TASK_RUNNING))
+	assert.Equal(state, StatePending)         // old state
+	assert.Equal(a_clone.State, StateRunning) // new state
+
+	// update unknown task
+	unknown_clone, state := registry.UpdateStatus(fakeStatusUpdate("unknown-task-id", mesos.TaskState_TASK_RUNNING))
+	assert.Nil(unknown_clone)
+	assert.Equal(state, StateUnknown)
+}
+
+func TestInMemoryRegistry_Update(t *testing.T) {
+	assert := assert.New(t)
+
+	// create offers registry
+	ttl := time.Second / 4
+	config := offers.RegistryConfig{
+		DeclineOffer: func(offerId string) <-chan error {
+			return proc.ErrorChan(nil)
+		},
+		Compat: func(o *mesos.Offer) bool {
+			return true
+		},
+		TTL:       ttl,
+		LingerTTL: 2 * ttl,
+	}
+	storage := offers.CreateRegistry(config)
+
+	// Add offer
+	offerId := mesosutil.NewOfferID("foo")
+	mesosOffer := &mesos.Offer{Id: offerId}
+	storage.Add([]*mesos.Offer{mesosOffer})
+	offer, ok := storage.Get(offerId.GetValue())
+	assert.True(ok)
+
+	// create registry
+	registry := NewInMemoryRegistry()
+	a, _ := fakePodTask("a")
+	registry.Register(a.Clone(), nil) // here clone a because we change it below
+
+	// state changes are ignored
+	a.State = StateRunning
+	err := registry.Update(a)
+	assert.NoError(err)
+	a_clone, _ := registry.Get(a.ID)
+	assert.Equal(StatePending, a_clone.State)
+
+	// offer is updated while pending
+	a.Offer = offer
+	err = registry.Update(a)
+	assert.NoError(err)
+	a_clone, _ = registry.Get(a.ID)
+	assert.Equal(offer.Id(), a_clone.Offer.Id())
+
+	// spec is updated while pending
+	a.Spec = Spec{SlaveID: "slave-1"}
+	err = registry.Update(a)
+	assert.NoError(err)
+	a_clone, _ = registry.Get(a.ID)
+	assert.Equal("slave-1", a_clone.Spec.SlaveID)
+
+	// flags are updated while pending
+	a.Flags[FlagType("launched")] = struct{}{}
```

Comment:
please use `Launched` const instead of `FlagType("launched")`, here throughout the remaining test(s)